### ISSUE TITLE
Add rename detection

### ISSUE THREAD

**Issue by martinvonz** [martinvonz/jj#47]
We should use rename detection at least when diffing and merging.

The footnote on https://blog.palantir.com/optimizing-gits-merge-machinery-2-d81391b97878 has some information about how Git does it.

**Comment by dbarnett** [comment]
A few logistical questions:
1. Is rename tracking explicit metadata (like in mercurial) or implicit (like git)? Do we know yet?
2. Are there baby steps towards getting this working, like for example updating `jj diff` to notice renames but nothing else?
3. Do we anticipate big gotchas with how rename tracking interacts with other features, like how the semantics are shared to git and other backends or breaking how conflicts work in renamed files?

I ran into an especially ugly rename+edit case recently and got intensely interested in improving this. =) I'd wanted to use `jj split` to separate revisions for the pure rename from the subsequent edits and that did not go well...

**Comment by joyously** [comment]
> Is rename tracking explicit metadata (like in mercurial) or implicit (like git)?

This question is confusing because Git doesn't deal with files (it has blobs), but it does have a `mv` command.
Bazaar(Breezy) also has a `mv` command, and it deals with renames better than Git.

**Comment by martinvonz** [comment]
We don't know yet if we're going to do rename tracking or rename detection. I'm leaning towards tracking, mostly because it seems like it would scale better. One interesting case that rename detection handles well when commit A adds file foo, commit B modifies it, and then you amend A with a rename from foo to bar (https://bz.mercurial-scm.org/show_bug.cgi?id=5457).

Also, if we want to track renames and also track content moves at the sub-file level (like `git blame -C`), then we would either have to track moved content at the sub-file level, or we'd have to somehow augment the file-level tracking information. It might be enough to track that "some content was moved out of file foo and into file bar" and then detect at query time exactly which content it was.

We should also keep in mind that diffing will hopefully get smarter in the future. In particular, I hope we'll start doing language-/AST-aware diffing and merging at some point. So we probably don't want to do any sub-file-level tracking of moved content, because we'll probably be able to detect exactly what moved in the future. Relevant message from Linus Torvalds: https://gist.github.com/borekb/3a548596ffd27ad6d948854751756a08

**Comment by martinvonz** [comment]
> > Is rename tracking explicit metadata (like in mercurial) or implicit (like git)?
> 
> This question is confusing because Git doesn't deal with files (it has blobs), but it does have a `mv` command. Bazaar(Breezy) also has a `mv` command, and it deals with renames better than Git.

Right, I think that's exactly what David is saying - that Git doesn't track renames. Mercurial does track renames.

**Comment by ChrisPenner** [comment]
The lack of any way to resolve conflicts where I've edited a file and I merge or rebase something which has renamed that file is a bit of a bummer at the moment;

I'd appreciate even a manual way to fix this, e.g. if I could edit the conflicted revision and somehow tell jj that the file isn't deleted it just moved.

**Comment by MultiCoreNop** [comment]
I'd like to mention another use case I stumbled upon recently: I modularized a project and (among other things) I've split a file into several pieces. (Imagine having struct `Foo` and struct `Bar` in `stuff.rs` and move `Foo` to `foo.rs` and `Bar` to `bar.rs`)
Meanwhile, a project member continued to work as usual while my restructuring was ongoing. After I was finished, I now had to rebase that project member's branch on top of my changes. Now it's not only a simple rename. If one could somehow 'target' a hunk to a specific file[+line], that would be a pretty neat feature, e.g. in the diff TUI.

The TUI might suggest all places, where it found the "base hunk" (of course, depending on the type of change and type of project, this could be very ambiguous and any form of automatism here might patch a completely wrong place).

**Comment by joyously** [comment]
> I now had to rebase that project member's branch on top of my changes.

Did you have to? Or would it work to rebase the other changes onto the parent of your changes (just before you split the files)?

**Comment by MultiCoreNop** [comment]
For sake of this example we can even assume, that my changes and the other ones actually shared the same parent from the beginning.
Yet, my PR passed first - so it then was "main" and therefore "the law" and everything else had to come later. (Imagine the restructuring going on for several weeks as well as the further development. At one point one has to merge something. It would probably be a similar high effort to "redo" the restructuring on the then updated main branch as it is to "re-integrate" the work that has been done in parallel into the restructured branch.)
Nevertheless, the merge itself is quite a brain-dead op - 90% simple copy and pasting, as restructuring did not touch the actual functionality of the code. So if one could direct `jj` to move "this patch hunk here " and "that one there", it would speed up things tremendously. Plain renames could of course be completely automatic, as they already are with Git (which I had to fall back to in this case, as it did merge 80% automatically what would otherwise have cost me days).

**Comment by martinvonz** [comment]
We don't seem to have a tracking issue about support for copies and renames in general. I think this is close but the title I picked specifically says "detection", which is not what the design doc describes. I'll change the title to make this one be our canonical issue for copies and renames.
